### PR TITLE
Solved desing problem with to_datetime when "ignore" error was being used.

### LINKED ISSUE #28120: to_datetime errors='ignore' unexpected behavior
I'm trying to apply pd.to_datetime on an integer column to convert integers (including missing indicators) to datetimes. In order to preserve all the missing indicators, ideally, I would prefer setting the "errors" parameter to "ignore" to keep all the invalid parsings. But then I noticed that setting errors='ignore' not only preserved the missing indicators, but kept everything the same, even the dtype of the input column. See below for more details.  
 
```python
print(dt_col)
```
<Outputs>
0    20181223

1    20161017
2      -99998
3    20171220
4    20170511
Name: dt_col, dtype: int64
</Outputs>

```python
pd.to_datetime(dt_col, format='%Y%m%d', errors='ignore')
```
<Outputs>
0    20181223

1    20161017
2      -99998
3    20171220
4    20170511
Name: dt_col, dtype: int64

</Outputs>

```python
pd.to_datetime(dt_col, format='%Y%m%d', errors='coerce')
```
<Outputs>
0   2018-12-23

1   2016-10-17
2          NaT
3   2017-12-20
4   2017-05-11
Name: dt_col, dtype: datetime64[ns]

</Outputs>

Pandas version: 
- pandas==0.25.1

Is anyone else seeing the same behavior?

### PR DESCRIPTION
**Pull request by IngErnestoAlvarez**
- [x] closes #28120
- [x] tests added / passed `test_to_datetime_some_invalid_numbers`
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] Now when u use some array of integers on to_datetime function passing as arguments: error="ignore" and format="%Y%m%d" it only skips none fitting values instead of all the values.

### CODE REVIEW

File: pandas/tests/tools/test_to_datetime.py [hunk 1]
```diff
@@ -71,6 +71,16 @@ def test_to_datetime_format(self, cache):
                 else:
                     tm.assert_index_equal(result, expected)
 
+    def test_to_datetime_format_some_invalid_numbers(self):
+        s = Series([20181223, 20161017, -9900, 20171220, 20170511])
```
Comment (arw2019) [inline pandas/tests/tools/test_to_datetime.py:75]: we try to avoid single letter variable names, better to use `ser`

File: pandas/tests/tools/test_to_datetime.py [hunk 2]
```diff
@@ -71,6 +71,16 @@ def test_to_datetime_format(self, cache):
...
+        expected = Series(['2018-12-23', '2016-10-17', -9900,
```
Comment (arw2019) [inline pandas/tests/tools/test_to_datetime.py:76]: I'd construct `expected` in a single step here - this is a little hard to parse IMO

File: pandas/tests/tools/test_to_datetime.py [hunk 3]
```diff
@@ -71,6 +71,16 @@ def test_to_datetime_format(self, cache):
...
+                           '2017-12-20', '2017-05-11'])
+        expected = Series([Timestamp(x) if isinstance(x, str) else x for x in expected])
+        result = to_datetime(s, format="%Y%m%d", errors='ignore')
+        print(result)
```
Comment (arw2019) [inline pandas/tests/tools/test_to_datetime.py:80]: we don't want print statements in the test (I'm guessing this is a mistake)

**Review by arw2019 — COMMENTED** [review]
Some comments on the test

But also looking at the issue it looks like opinion is divided on whether we want to do this

**Comment by IngErnestoAlvarez** [comment]
@arw2019  Hi! I leave the code here, with the changes you recommend me to do. I thought it was an open issue, not necessarily a bug, but a feature that can be good for it to be used. What I would understand at first, seeing this, `to_datetime(dt_col, format='%Y%m%d', errors='ignore')`, is that it will transform the array to datetime omitting all the errors and exceptions. It doesn't sound like "if I found an error I won't change anything". But I understand if you see it differently. Thank you for the feedback!

**Review by arw2019 — CHANGES_REQUESTED** [review]
you need to run black to get CI passing - then others will re-review

File: pandas/_libs/tslibs/strptime.pyx [hunk 4]
```diff
@@ -141,12 +145,20 @@ def array_strptime(ndarray[object] values, object fmt, bint exact=True, errors='
                 if is_coerce:
                     iresult[i] = NPY_NAT
                     continue
+                if is_ignore:
```
Comment (jreback) [inline pandas/_libs/tslibs/strptime.pyx:148]: can be elif

File: pandas/_libs/tslibs/strptime.pyx [hunk 5]
```diff
@@ -141,12 +145,20 @@ def array_strptime(ndarray[object] values, object fmt, bint exact=True, errors='
...
+                    new_result[i] = int(val)
+                    return_object_array = True
+                    continue
                 raise ValueError(f"time data '{val}' does not match "
                                  f"format '{fmt}' (match)")
             if len(val) != found.end():
                 if is_coerce:
                     iresult[i] = NPY_NAT
                     continue
+                if is_ignore:
```
Comment (jreback) [inline pandas/_libs/tslibs/strptime.pyx:158]: same

File: pandas/_libs/tslibs/strptime.pyx [hunk 6]
```diff
@@ -156,6 +168,10 @@ def array_strptime(ndarray[object] values, object fmt, bint exact=True, errors='
...
```
Comment (jreback) [inline pandas/_libs/tslibs/strptime.pyx:171]: same

File: pandas/_libs/tslibs/strptime.pyx [hunk 7]
```diff
@@ -315,6 +331,10 @@ def array_strptime(ndarray[object] values, object fmt, bint exact=True, errors='
             if is_coerce:
                 iresult[i] = NPY_NAT
                 continue
+            if is_ignore:
```
Comment (jreback) [inline pandas/_libs/tslibs/strptime.pyx:334]: same

File: pandas/_libs/tslibs/strptime.pyx [hunk 8]
```diff
@@ -335,10 +355,16 @@ def array_strptime(ndarray[object] values, object fmt, bint exact=True, errors='
...
```
Comment (jreback) [inline pandas/_libs/tslibs/strptime.pyx:358]: same

File: pandas/tests/tools/test_to_datetime.py [hunk 9]
```diff
@@ -71,6 +71,20 @@ def test_to_datetime_format(self, cache):
                 else:
                     tm.assert_index_equal(result, expected)
 
+    def test_to_datetime_format_some_invalid_numbers(self):
```
Comment (jreback) [inline pandas/tests/tools/test_to_datetime.py:74]: you added a ton of conditionals, i would expect that you can parameterize so that each is exercised (or simply not needed)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/tslibs/strptime.pyx [hunk 10]
```diff
@@ -118,6 +121,7 @@ def array_strptime(ndarray[object] values, object fmt, bint exact=True, errors='
     result = np.empty(n, dtype='M8[ns]')
     iresult = result.view('i8')
     result_timezone = np.empty(n, dtype='object')
+    new_result = np.empty(n, dtype='O8')
```
Comment (jreback) [inline pandas/_libs/tslibs/strptime.pyx:124]: also i am a bit concerned that this is going to be very heavyweight (e.g. this is going to be 10x memory as an object array is much heavier here). so even though this is a bit more complicated, i would put the main loop in a try/except and then if it excepts then do the object dtype conversion (only). now I don't entirely remember this code so this may blow up complexity but it is worth thinking about.

this pattern we do elsewhere (though @jbrockmendel maybe we removed / reduced it?)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/tslibs/strptime.pyx (hunk 10 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/strptime.pyx:124]: yah thats the pattern we use in array_to_datetime

also why 'O8' instead of just 'O'?

Comment (IngErnestoAlvarez) [reply pandas/_libs/tslibs/strptime.pyx:124]: Do you suggest that is better to make the main loop twice if an exception was thrown than having a second array for the main loop lifetime? Okay, that can be possible. I will check `array_to_datetime` and see how did it was done.

Comment (jreback) [reply pandas/_libs/tslibs/strptime.pyx:124]: yep exactly

Comment (IngErnestoAlvarez) [reply pandas/_libs/tslibs/strptime.pyx:124]: In respect to "O" or "O8", I thought I needed to maintain the dtype size, if before was an "I8" and transformed to object type it had to be in 8 bytes too.

Comment (jreback) [reply pandas/_libs/tslibs/strptime.pyx:124]: O is just fine (or `dtype=object`

Comment (IngErnestoAlvarez) [reply pandas/_libs/tslibs/strptime.pyx]: I just added the new function just like u said. Now I created `array_strptime_object` that makes the same as `array_strptime` but returns the array with object dtype. I ran all the tests and all seems to work fine.
I did exactly the same as the function u gave to me. `array_to_datetime_object`

File: pandas/_libs/tslibs/strptime.pyx [hunk 11]
```diff
@@ -754,3 +775,232 @@ cdef tzinfo parse_timezone_directive(str z):
...
+            if len(val) != found.end():
+                result[i] = int(val)
+
+        # search
+        else:
+            found = format_regex.search(val)
+            if not found:
+                result[i] = int(val)
+
+        iso_year = -1
```
Comment (jreback) [inline pandas/_libs/tslibs/strptime.pyx:832]: we are going to want to factor this out into a function instead of repeating this for both

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by IngErnestoAlvarez** [comment]
I'm not done yet, but It seems that in one Travis build conda had an error "solving the environment", it just gets stuck, it's something I should be worried about? Or first I finish what I'm doing and then take care of that?

**Comment by jreback** [comment]
ignore travis as it's being worked elsewhere (though certainly would appreciate help in that - we r trying to move off it)

**Comment by IngErnestoAlvarez** [comment]
What u r trying to do? Changing the CI environment? 
What things do u need help with? Maybe I can help

**Comment by jreback** [comment]
so this build: https://travis-ci.org/github/pandas-dev/pandas/jobs/751146161

something changed very recently like in last day or so - and now conda solve time blows up

so we want to fix the install env for just this build to make it work

independent we are working on moving these builds to azure / but that's slightly orthongoal

**Comment by jreback** [comment]
closing as stale. if you want to continue, pls ping and can re-open.
